### ISSUE TITLE
Add ConvBrain to GAN and WGAN

### ISSUE THREAD

**Issue by wvabrinskas** [wvabrinskas/Neuron#18]
- Add support for `ConvBrain`s in `WGAN` and `GAN` classes. 
- Right now it is expecting Brain objects. This is not how GANs should be created. 
- Updating this will require some refactoring...

**Comment by wvabrinskas** [comment]
It looks like we'll need to completely refactor the way `ConvBrain` and `Brain` work. 

https://machinelearningmastery.com/how-to-code-a-wasserstein-generative-adversarial-network-wgan-from-scratch/

- You can see there that a `Fully Connected` layer is first THEN a `Conv` layer with a `Reshape` layer in between.
- We will need to rework how a `Brain` object is created and consolidate everything into one class that can do both with any arbitrary array dimension. 

Converting between dimensions is the easy part. Supporting arbitrary array dimensions will be a bit trickier.

**Comment by wvabrinskas** [comment]
For the `Generator` we'll need to add a new layer called the `ConvolutionTransposed` layer. This will enlarge the input by X dimensions via the padding / stride option. 

https://datascience.stackexchange.com/questions/6107/what-are-deconvolutional-layers

**Comment by wvabrinskas** [comment]
Google Quick Draw dataset: https://console.cloud.google.com/storage/browser/quickdraw_dataset/full/numpy_bitmap;tab=objects?pli=1&prefix=&forceOnObjectsSortingFiltering=false

**Comment by wvabrinskas** [comment]
Back propagation with transposed layer: https://stats.stackexchange.com/questions/420043/backpropagation-through-2d-transposed-convolution-layer

**Comment by wvabrinskas** [comment]
https://naokishibuya.medium.com/up-sampling-with-transposed-convolution-9ae4f2df52d0

**Comment by wvabrinskas** [comment]
This was incredibly helpful with debugging: https://medium.com/ai-society/gans-from-scratch-1-a-deep-introduction-with-code-in-pytorch-and-tensorflow-cb03cdcdba0f

**Comment by wvabrinskas** [comment]
https://hideyukiinada.github.io/cnn_backprop_strides2.html Helped with understanding backpropagation with strides

**Comment by wvabrinskas** [comment]
https://medium.com/@mayank.utexas/backpropagation-for-convolution-with-strides-8137e4fc2710
https://medium.com/@mayank.utexas/backpropagation-for-convolution-with-strides-fb2f2efc4faa

**Comment by wvabrinskas** [comment]
https://www.tensorflow.org/tutorials/generative/dcgan

**Comment by wvabrinskas** [comment]
https://github.com/palle-k/DL4S-WGAN-GP This gave me an idea about how to make this more concurrent
